modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/data_renderers/DateRenderer.js

```diff
@@ -14,18 +14,20 @@
 
 import PropType from 'prop-types';
 
+import {getDateFromDateString, prettifyDateObject} from '../utils/dates';
+
 function DateRenderer({options, value}) {
 	if (!value) {
 		return null;
 	}
 
+	const dateObject = getDateFromDateString(value.split('T')[0]);
+	value = prettifyDateObject(dateObject);
```

Comment:
Hey @markocikos , thanks for your detailed comments.
I'm not 100% sure, but I think ISO 8601 format is the default pattern to represent a Date in JSON payload of headless services on Liferay. Maybe @javierdearcos can help us about it.

Let me explain how this problem arise in Objects, for now we just have Date field type (we will also have Date Time type soon), so when some Object has this field type and user entries a value, we have a Date without the time part / with default value 00:00:00.000
The Taglib clay:headless-data-set-display uses Rest API to fetch and display data, we don't have access to transform the values, we are just able to specify the fields implementing BaseTableClayDataSetDisplayView like in ObjectEntriesTableClayDataSetDisplayView, so this Date returned by API comes in ISO format like "2021-12-09T00:00:00Z", this value is currently shown in portlets, but we need to show it in a friendly format. 
First we tried to set clayTableSchemaField.setContentRenderer("date"), but DateRenderer.js logic applies time zone offset in local time when format the value, consequently the Date is displayed one day before (2021-12-08) for Brazil time zone (GMT/UTC - 03:00), behind the scenes it becomes "2021-12-08 21:00:00"

Our idea is to have a way to inform to the Taglib that for some fields (ClayTableSchemaField) we are displaying a Date (ContentRenderer("date")) and for others a Date Time ContentRenderer("datetime"). There are these 2 use cases.
After some searching on Google and doing some tests we got the desired behavior using value.split('T')[0]

Please let us know if you have any doubt about the problem or if you have another solution for that.

Reply:
> First we tried to set clayTableSchemaField.setContentRenderer("date"), but DateRenderer.js logic applies time zone offset in local time when format the value, consequently the Date is displayed one day before (2021-12-08) for Brazil time zone (GMT/UTC - 03:00), behind the scenes it becomes "2021-12-08 21:00:00"

The problem is, I don't think there is no good way in JS to get around this. The current solution we have is very hacky, part insane and part genius:
1. First step: we convert `yyyy-mm-dd` to `mm/dd/yyyy`. This is basically what  `getDateFromDateString` + `prettifyDateObject` do. For this, we rely on [date.toLocalDateString](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/dates.js#L38), where locale is not set.
2. Second step: we rely on [Date parsing](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/data_renderers/DateRenderer.js#L32) to resolve `mm/dd/yyyy` as UTC midnight.
3. Third step: convert UTC date to local format, with `Intl.DateTimeFormat` and its options. This step is fine.

First two steps use very counterintuitive JS behavior, as well as being flaky and browser inconsistent. I sincerely doubt that all browsers in all timezones return `mm/dd/yyyy` as default format, and we know from MDN that Date parsing is browser inconsistent.

Reply:
Hi @markocikos, to avoid creating dates with strings, what I've managed to do here is extract the object number elements that come from the `getDateFromDateString function (value.split('T') [0]);` and pass them as arguments when creating the date, without having to go through the `prettifyDateObject (dateObject)` function. Like this:

```
	const {year, month, day} = getDateFromDateString(value.split('T')[0]);

	const locale = themeDisplay.getBCP47LanguageId();
	const dateOptions = options?.format || {
		day: 'numeric',
		month: 'short',
		year: 'numeric',
	};

	const formattedDate = new Intl.DateTimeFormat(locale, dateOptions).format(
		new Date(year, month - 1, day)
	);

```
For some reason, if I just pass the month the `Intl.DateTimeFormat` function adds another month (I don't really know why :laughing: :thinking: ), so to get around that I have to subtract one from the month. I want to know what your opinion about this?

Reply:
Hey @danvb112 , I think it's ok, as monthIndex starts with 0 (January), I saw this on [MDN link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date) provided above. Am I right @markocikos ?

> monthIndex
> Integer value representing the month, beginning with 0 for January to 11 for December. If a value greater than 11 is passed in, then those months will be added to the date; for example, new Date(1990, 12, 1) will return January 1st, 1991

Reply:
@danvb112 Your solution in https://github.com/liferay-frontend/liferay-portal/pull/1722#discussion_r766948558 definitely an improvement, as we are no longer using the inconsistent `Date` parsing. What I don't like is that it is still a bit hacky: we are representing date that was originally a UTC date with an equivalent date in user's timezone. While it does work for displaying a date that matches current inputs we have, it might be a trap for anybody extending this code or API in the future. For example, if why try to explicitly interpret input as UTC, we get the wrong date: 
```js
new Date(2021, 11, 01).getTime() // 1638313200000

new Intl.DateTimeFormat('en-GB').format(new Date(2021, 11, 01)) // '01/12/2021'

Date.UTC(2021, 11, 01) // 1638316800000

new Intl.DateTimeFormat('en-GB', {timeZone: 'UTC'}).format(new Date(2021, 11, 01)) // '30/11/2021'
```

What we can do it be explicitly use UTC in both parsing and final output. Something like:

```js
let timestamp = value;

if (typeof value === 'string') {
	const date = value.split('T')[0];

	const dateArray = date.split('-');

	if (dateArray.length === 3) {
		const [year, month, day] = dateArray;

		timestamp = Date.UTC(Number(year), Number(month) - 1, Number(day));
	}
	else {
		timestamp = Number(value);
	}
}

const dateOptions = {
	day: options?.format?.day || 'numeric',
	month: options?.format?.month || 'short',
	timeZone: options?.format?.timezone || 'UTC',
	year: options?.format?.year || 'numeric',
};

const formattedDate = new Intl.DateTimeFormat(locale, dateOptions).format(
	timestamp
);
```

As a bonus, we are enabling timestamp input. 

What do you think?

Reply:
I understand your point @markocikos, I think your suggestion is very good, I'll do it now, thanks :smiley: